### PR TITLE
Merge `Branch` type stubs back to source

### PR DESCRIPTION
**Pull request by trim21**

### CODE REVIEW

File: github/Branch.py
Comment (trim21) [reply github/Branch.py:165]: Are you going to merge https://github.com/PyGithub/PyGithub/pull/2535 first or merge this PR first? If you are going to merge 2535 first, I'll wait after it's merged to avoid extra conflect.

File: github/Branch.py
Comment (JLLeitschuh) [reply github/Branch.py:133]: Oh, interesting! Any documentation on it being deprecated?

Comment (trim21) [reply github/Branch.py:133]: https://docs.python.org/3/library/typing.html#typing.List

File: github/Branch.py
Comment (EnricoMi) [reply github/Branch.py:163]: Merged.

File: github/Branch.py
Comment (trim21) [inline github/Branch.py:173]: @EnricoMi is this okay?

File: github/Branch.py
Comment (trim21) [inline github/Branch.py:220]: minor refactor here, avoid `... or ... or ... or ... or ... or ...`

**Comment by trim21** [comment]
@EnricoMi why don't we just use `None` as default value here?

```py
        if is_defined(lock_branch):
            post_parameters["lock_branch"] = lock_branch
        else:
            post_parameters["lock_branch"] = None

        if is_defined(allow_fork_syncing):
            post_parameters["allow_fork_syncing"] = allow_fork_syncing
        else:
            post_parameters["allow_fork_syncing"] = None

        if is_defined(block_creations):
            post_parameters["block_creations"] = block_creations
        else:
            post_parameters["block_creations"] = None
```

File: github/GithubObject.py
Comment (JLLeitschuh) [inline github/GithubObject.py:99]: Why is this `TypeGuard[List[T]]` vs `TypeGuard[Optional[T]]`?

File: github/GithubObject.py
Comment (JLLeitschuh) [inline github/GithubObject.py:108]: I like these! Good idea!

**Review by JLLeitschuh — COMMENTED** [review]
Some questions/comments. This looks solid though! Nice work!!!

File: github/GithubObject.py
Comment (trim21) [reply github/GithubObject.py:99]: Oh, this is wrong... it's `TypeGuard[Opt[T]]`

File: github/GithubObject.py
Comment (EnricoMi) [inline github/GithubObject.py:96]: What about
```suggestion
def is_set(v: Union[T, _NotSetType]) -> TypeGuard[T]:
```

File: github/GithubObject.py
Comment (EnricoMi) [inline github/GithubObject.py:100]: what about
```suggestion
def is_not_set(v: Any) -> TypeGuard[_NotSetType]:
```
given default values are `NotSet`.

File: github/GithubObject.py
Comment (EnricoMi) [inline github/GithubObject.py:108]:
```suggestion
def is_optional_list(v, type: Type[T]) -> TypeGuard[List[T]]:
```

Comment (EnricoMi) [reply github/GithubObject.py:108]: Shouldn't return type be
```suggestion
def is_optional_list_of_type(v, type: Type[T]) -> TypeGuard[Opt[List[T]]]:
```

**Comment by EnricoMi** [comment]
> @EnricoMi why don't we just use `None` as default value here?

In these particular cases, that looks like the right thing to do.

File: github/Branch.py
Comment (EnricoMi) [reply github/Branch.py]: This touches a lot of logic, we'll have to carefully review this.

File: github/GithubObject.py
Comment (trim21) [reply github/GithubObject.py:96]: `set` is a python builtin type, maybe confusing.

File: github/Branch.py
Comment (trim21) [reply github/Branch.py]: I can revert them to previous logic, only change `... is NotSet`

File: github/GithubObject.py
Comment (EnricoMi) [reply github/GithubObject.py:96]: Well, this holds for `NotSet` and `_NotSetType` as well.

Let's stick to `defined` then.

File: github/Branch.py
Comment (EnricoMi) [reply github/Branch.py]: Lets not rewrite the logic, stick to merging typing into source, which only includes the `NotSet` checking. The rest of the improvements should be done in a separate PR (e.g. `None` defaults).

File: github/GithubObject.py
Comment (trim21) [reply github/GithubObject.py:96]: maybe rename `NotSet` to `Undefined`

Comment (EnricoMi) [reply github/GithubObject.py:96]: Like that, but in a separate PR, will touch all files :-/

**Review by EnricoMi — APPROVED** [review]
Code becomes so much more readable with these new helper methods. LGTM!

**Pull request merged**
